[CELEBORN-1122] Metrics supports json format

suizhe007 opened the pull request:
<!--
Thanks for sending a pull request!  Here are some tips for you:
  - Make sure the PR title start w/ a JIRA ticket, e.g. '[CELEBORN-XXXX] Your PR title ...'.
  - Be sure to keep the PR description updated to reflect all changes.
  - Please write your PR title to summarize what this PR proposes.
  - If possible, provide a concise example to reproduce the issue for a faster review.
-->

### What changes were proposed in this pull request?
If the user does not use prometheus to collect monitoring metrics, but rather some other ones. Using metrics in JSON format would be more user-friendly.The PR supports JSON format for metrics.


### Why are the changes needed?
Ditto.


### Does this PR introduce _any_ user-facing change?
Metrics supports JSON format


### How was this patch tested?
Cluster test.

RexXiong commented on common/src/main/scala/org/apache/celeborn/common/metrics/sink/JsonServlet.scala:
```diff
@@ -0,0 +1,316 @@
+    val metricDatas = new util.ArrayList[MetricData]
+    innerMetrics.synchronized {
+      while (!innerMetrics.isEmpty) {
+        metricDatas.add(innerMetrics.poll)
```
When getMetrics concurrent(though its rare), innerMetrics may be record duplicate metrics, May be we can cache snapshot metrics data(for a while), instead use innerMetrics and synchronized it for every request

suizhe007 replied:
> When getMetrics concurrent(though its rare), innerMetrics may be record duplicate metrics, May be we can cache snapshot metrics data(for a while), instead use innerMetrics and synchronized it for every request

I think it makes sense , and make some corrections.

suizhe007 replied:
Thank you for your suggestion. I think it makes sense , and make some corrections.

> When getMetrics concurrent(though its rare), innerMetrics may be record duplicate metrics, May be we can cache snapshot metrics data(for a while), instead use innerMetrics and synchronized it for every request

@RexXiong  Thank you for your suggestion. I think it makes sense , and make some corrections.

RexXiong commented on common/src/main/scala/org/apache/celeborn/common/metrics/sink/JsonServlet.scala:
```diff
@@ -0,0 +1,316 @@
+
+  private def updateInnerMetrics(absSource: AbstractSource, metricData: MetricData): Unit = {
+    innerMetrics.synchronized {
+      if (innerMetrics.size() >= absSource.metricsCapacity) {
```
if condition meet, can we directly return? as remove and add new metricData seems useless

suizhe007 replied:
> if condition meet, can we directly return? as remove and add new metricData seems useless

Thank you for your suggestion. I have resubmitted, please help me review it again.

suizhe007 replied:
> if condition meet, can we directly return? as remove and add new metricData seems useless

@RexXiong   Thank you for your suggestion. I have resubmitted, please help me review it again.

suizhe007 commented on common/src/main/scala/org/apache/celeborn/common/metrics/MetricsSystem.scala:
```diff
@@ -125,15 +129,27 @@ class MetricsSystem(
+                sources.asScala,
+                jsonServletPath).asInstanceOf[JsonServlet])
+            }
+            if (kv._1 == "prometheusServlet") {
```
> nit: else if

@RexXiong  Thank you for your suggestion.

RexXiong commented on common/src/main/scala/org/apache/celeborn/common/metrics/sink/JsonServlet.scala:
```diff
@@ -0,0 +1,337 @@
+
+import com.codahale.metrics.MetricRegistry
+import io.netty.channel.ChannelHandler.Sharable
+import org.apache.hadoop.shaded.com.fasterxml.jackson.annotation.JsonAutoDetect.Visibility
```
I have no other comments on this pr except this import as it from shaded dependency, WDYT @waitinfuture @pan3793 @FMX @cfmcgrady

suizhe007 replied:
> I have no other comments on this pr except this import as it from shaded dependency, WDYT @waitinfuture @pan3793 @FMX @cfmcgrady

@RexXiong   Thank you for your suggestion. I reimport the jackson package,and the version from hadoop 3.2.4.

pan3793 commented on pom.xml:
```diff
@@ -96,6 +96,8 @@
     <zstd-jni.version>1.5.2-1</zstd-jni.version>
     <kubernetes-client.version>6.7.0</kubernetes-client.version>
     <rocksdbjni.version>8.5.3</rocksdbjni.version>
+    <jackson.version>2.10.5</jackson.version>
```
why choose such a low version

suizhe007 replied:
> why choose such a low version
The version of `org. apache. hadoop. shaded. *. jackson` comes from Hadoop 3.2.4 now. In order to reduce its impact on the original framework. I directly import the jackson 2.5.1 from Hadoop 3.2.4.

pan3793 commented:
please update the LICENSE and NOTICE since it introduces the new dependencies, also, fix sbt building system

suizhe007 commented:
> please update the LICENSE and NOTICE since it introduces the new dependencies, also, fix sbt building system

@pan3793  Thank you. I have updated newer version , and fixed sbt building system. Please help me review it again.

pan3793 commented on project/CelebornBuild.scala:
```diff
@@ -56,6 +56,8 @@ object Dependencies {
   val roaringBitmapVersion = "0.9.32"
   val rocksdbJniVersion = "8.5.3"
+  val jacksonVersion = "2.15.3"
+  val paranamerVersion = "2.8"
```
seems we don't need to manage `paranamerVersion` explicitly, let's simply delegate it to jackson

suizhe007 replied:
> seems we don't need to manage `paranamerVersion` explicitly, let's simply delegate it to jackson

Thank you. I remove it.

pan3793 commented on common/src/main/scala/org/apache/celeborn/common/metrics/sink/JsonServlet.scala:
```diff
@@ -0,0 +1,346 @@
+import org.apache.celeborn.common.metrics.source.{AbstractSource, NamedCounter, NamedGauge, NamedHistogram, NamedTimer, Source}
+
+object JsonConverter {
+  val mapper = new ObjectMapper() with ClassTagExtensions
```
I guess user may preview the json using browser, maybe we can print the pretty json string.

pan3793 commented on service/src/main/scala/org/apache/celeborn/server/common/http/HttpRequestHandler.scala:
```diff
@@ -49,8 +49,14 @@ class HttpRequestHandler(
+          val servletHttpRequestHandler = servletHttpRequestHandlers.find(servlet =>
+            uri == servlet.getServletPath())
+          if (servletHttpRequestHandler.nonEmpty) {
+            servletHttpRequestHandler.get.handleRequest(uri)
```
should we update it for JSON?
```
res.headers().set(HttpHeaderNames.CONTENT_TYPE, "text/plain; charset=UTF-8")
```

suizhe007 replied:
> should we update it for JSON?
> 
> ```
> res.headers().set(HttpHeaderNames.CONTENT_TYPE, "text/plain; charset=UTF-8")
> ```

Thank you. You are right. I update it.

cfmcgrady commented on common/src/main/scala/org/apache/celeborn/common/metrics/MetricsSystem.scala:
```diff
@@ -27,30 +27,34 @@ import scala.util.matching.Regex
   private val sinks = new ArrayBuffer[Sink]
   private val sources = new CopyOnWriteArrayList[Source]
   private val registry = new MetricRegistry()
+  val prometheusServletPath = conf.get(METRICS_PROMETHEUS_PATH)
+  val jsonServletPath = conf.get(METRICS_JSON_PATH)
```
```suggestion
  private val prometheusServletPath = conf.get(METRICS_PROMETHEUS_PATH)
  private val jsonServletPath = conf.get(METRICS_JSON_PATH)
```

pan3793 commented on common/src/main/scala/org/apache/celeborn/common/metrics/sink/JsonServlet.scala:
```diff
@@ -0,0 +1,355 @@
+    mapper.writeValueAsString(value)
+  }
+
+  def toJsonWithPretty(value: ArrayBuffer[Any]): String = {
```
```suggestion
  def toPrettyJson(value: Any): String = {
```

suizhe007 replied:
Thank you. I update it.

pan3793 commented on LICENSE-binary:
```diff
@@ -280,6 +280,10 @@ org.scala-lang:scala-reflect
+com.fasterxml.jackson.module:jackson-module-scala
+com.fasterxml.jackson.core:jackson-databind
+com.fasterxml.jackson.core:jackson-annotations
+com.fasterxml.jackson.core:jackson-core
```
LICENSE-binary reflects the final content of the binary release tarball, so the parameter should be retained here.

suizhe007 replied:
> LICENSE-binary reflects the final content of the binary release tarball, so the parameter should be retained here.

what is `the parameter`?

pan3793 replied:
sorry, the paranamer

pan3793 commented on dev/deps/dependencies-client-spark-3.2:
```diff
@@ -20,10 +20,14 @@ commons-crypto/1.0.0//commons-crypto-1.0.0.jar
-hadoop-client-api/3.3.6//hadoop-client-api-3.3.6.jar
-hadoop-client-runtime/3.3.6//hadoop-client-runtime-3.3.6.jar
+guava/14.0.1//guava-14.0.1.jar
+hadoop-client-api/3.2.4//hadoop-client-api-3.2.4.jar
```
please rebase master and regenerate this files

suizhe007 replied:
Thank you. I update it.

suizhe007 commented on common/src/main/scala/org/apache/celeborn/common/metrics/sink/JsonServlet.scala:
```diff
@@ -0,0 +1,355 @@
+  val mapper = new ObjectMapper() with ClassTagExtensions
+  mapper.registerModule(DefaultScalaModule)
+
+  def toJson(value: ArrayBuffer[Any]): String = {
```
Thank you. I update it.

RexXiong reviewed (approved):
LGTM, cc @pan3793 @cfmcgrady

pan3793 commented on service/src/main/scala/org/apache/celeborn/server/common/http/HttpRequestHandler.scala:
```diff
@@ -47,21 +47,34 @@ class HttpRequestHandler(
     val msg = HttpUtils.handleRequest(service, path, parameters)
-    val response = msg match {
+    val textType = "text/plain; charset=UTF-8"
+    val jsonType = "application/json; charset=UTF-8"
```
```suggestion
    val jsonType = "application/json"
```

https://www.rfc-editor.org/rfc/rfc8259.html

>    Note:  No "charset" parameter is defined for this registration.
>    Adding one really has no effect on compliant recipients.

suizhe007 replied:
Thanks. I update it.

pan3793 reviewed (approved):
LGTM, except for one minor thing

cfmcgrady commented on service/src/main/scala/org/apache/celeborn/server/common/http/HttpRequestHandler.scala:
```diff
@@ -47,21 +47,34 @@ class HttpRequestHandler(
-        if (prometheusHttpRequestHandler != null) {
-          prometheusHttpRequestHandler.handleRequest(uri)
+        if (servletHttpRequestHandlers != null) {
+          val servletHttpRequestHandler = servletHttpRequestHandlers.find(servlet =>
+            uri == servlet.getServletPath())
+          servletHttpRequestHandler match {
+            case Some(handler) => handler match {
+                case jsonHandler: JsonHttpRequestHandler =>
+                  (jsonHandler.handleRequest(uri), jsonType)
+                case _ => (handler.handleRequest(uri), textType)
+              }
+            case _ => (s"Unknown path $uri!", textType)
+          }
```
nit
```suggestion
          servletHttpRequestHandlers.find(servlet => uri == servlet.getServletPath())
            .map {
              case jsonHandler: JsonHttpRequestHandler =>
                (jsonHandler.handleRequest(uri), jsonType)
              case handler: ServletHttpRequestHandler =>
                (handler.handleRequest(uri), textType)
            }.getOrElse((s"Unknown path $uri!", textType))
```

suizhe007 replied:
Thanks. I update it.

cfmcgrady reviewed (approved):
LGTM, except a nit

RexXiong commented:
Thank you all @suizhe007 @pan3793 @cfmcgrady, merge to main(v0.4.0)
